devito/ir/equations/algorithms.py

```diff
@@ -162,74 +163,129 @@ def concretize_subdims(exprs, **kwargs):
 
-@_concretize_subdims.register(ConditionalDimension)
-def _(d, mapper, sregistry):
-    # TODO: to be implemented as soon as we drop the counter machinery in
-    # Grid.__subdomain_finalize__
-    # TODO: call `_concretize_subdims(d.parent, mapper)` as the parent might be
-    # a SubDimension!
-    pass
+    mapper[d] = d._rebuild(symbolic_min=left, symbolic_max=right, thickness=thickness)
 
 
-@_concretize_subdims.register(MultiSubDimension)
-def _(d, mapper, sregistry):
-    if not d.is_abstract:
-        # TODO: for now Grid.__subdomain_finalize__ creates the thickness, but
-        # soon it will be done here instead
+@_concretize_subdims.register(ConditionalDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
+        # Already have a substitution for this dimension
         return
 
-    pd = d.parent
+    _concretize_subdims(d.parent, mapper, rebuilt, sregistry)
+
+    kwargs = {}
+
+    # Parent may be a subdimension
+    if d.parent in mapper:
+        kwargs['parent'] = mapper[d.parent]
+
+    # Condition may contain subdimensions
+    if d.condition is not None:
+        for v in d.condition.free_symbols:
+            _concretize_subdims(v, mapper, rebuilt, sregistry)
 
-    subs = mapper[pd]
+        if any(v in mapper for v in d.condition.free_symbols):
+            # Substitute into condition
+            kwargs['condition'] = d.condition.subs(mapper)
 
-    if d in subs:
+    if kwargs:
+        # Rebuild if parent or condition need replacing
+        mapper[d] = d._rebuild(**kwargs)
+
+
+@_concretize_subdims.register(MultiSubDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
         # Already have a substitution for this dimension
         return
 
-    name = sregistry.make_name(prefix=d.name)
-    ltkn, rtkn = MultiSubDimension._symbolic_thickness(name)
+    abstract_tkns = MultiSubDimension._symbolic_thickness(d.parent.name)
+    concrete_tkns = tuple(tkn._rebuild(name=sregistry.make_name(prefix=tkn.name))
+                          for tkn in abstract_tkns)
 
-    thickness = (ltkn, rtkn)
+    kwargs = {'thickness': concrete_tkns}
+    fkwargs = {}
 
-    subs[d] = d._rebuild(d.name, pd, thickness)
+    idim0 = d.implicit_dimension
+    if idim0 is not None:
+        try:
+            idim1 = rebuilt[idim0]
+        except KeyError:
+            iname = sregistry.make_name(prefix=idim0.name)
+            rebuilt[idim0] = idim1 = idim0._rebuild(name=iname)
+
+        kwargs['implicit_dimension'] = idim1
+        fkwargs.update({'dimensions': (idim1,) + (d.functions.dimensions[1:]),
+                        'halo': None,
+                        'padding': None})
+
+    try:
+        functions = rebuilt[d.functions]
+    except KeyError:
+        fkwargs.update({'name': sregistry.make_name(prefix=d.functions.name),
+                        'function': None,
+                        'allocator': ExternalAllocator(d.functions.data),
```

Comment:
`._data`

`.data` instead is just a view. Suggests that you need a test with space_order != 0 to trigger what would be an issue if you don't replace `.data` with `._data`